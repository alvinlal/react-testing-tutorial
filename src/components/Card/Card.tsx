import { useState } from 'react';
import { ReactComponent as HeartFilled } from '../../assets/icons/heartFilled.svg';
import { ReactComponent as HeartOutlined } from '../../assets/icons/heartOutlined.svg';
import './Card.css';

const Card: React.FC<Cat> = ({ name, phone, email, image: { url, alt }, favoured }) => {
  const [isFavoured, setIsFavoured] = useState<boolean>(favoured);

  return (
    <div className='card' data-testid='cat-card'>
      <div className='card-header'>
        <img src={url} alt={alt} className='card-img' />
        <button className='heart' onClick={() => setIsFavoured(!isFavoured)}>
          {isFavoured ? (
            <HeartFilled data-testid='filled-heart' />
          ) : (
            <HeartOutlined data-testid='outlined-heart' />
          )}
        </button>
        <div className='card-content'>
          <h3>{name}</h3>
          <p>{phone}</p>
          <p>{email}</p>
        </div>
      </div>
    </div>
  );
};

export default Card;
