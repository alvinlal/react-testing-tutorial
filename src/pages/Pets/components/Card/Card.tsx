import { useState } from 'react';
import { ReactComponent as HeartFilled } from '../../../../assets/icons/heartFilled.svg';
import { ReactComponent as HeartOutlined } from '../../../../assets/icons/heartOutlined.svg';
import './Card.css';

interface CardProps extends Cat {
  index: number;
  updateFavoured: (index: number, favoured: boolean) => void;
}

const Card: React.FC<CardProps> = ({
  index,
  name,
  phone,
  email,
  image: { url, alt },
  favoured,
  updateFavoured,
}) => {
  const [isFavoured, setIsFavoured] = useState<boolean>(favoured);

  const toggleFavoured = () => {
    updateFavoured(index, !isFavoured);
    setIsFavoured(!isFavoured);
  };

  return (
    <div className='card' data-testid='cat-card'>
      <div className='card-header'>
        <img src={url} alt={alt} className='card-img' />
        <button className='heart' onClick={toggleFavoured}>
          {isFavoured ? (
            <HeartFilled data-testid='filled-heart' />
          ) : (
            <HeartOutlined data-testid='outlined-heart' />
          )}
        </button>
        <div className='card-content'>
          <h3>{name}</h3>
          <p>{phone}</p>
          <p>{email}</p>
        </div>
      </div>
    </div>
  );
};

export default Card;
