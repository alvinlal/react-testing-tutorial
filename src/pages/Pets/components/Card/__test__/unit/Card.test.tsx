/**
 * Requirements
 * 1. Should display the name of the cat
 * 2. Should display the phone number of the owner
 * 3. Should display the email of the owner
 * 4. Should show the image of the cat with correct src
 * 5. Should show outlined heart
 * 6. Should show filled heart
 * 7. Should toggle heart status
 */

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Card from '../../Card';

describe('Card.tsx', () => {
  const cat: Cat = {
    id: 1,
    name: 'Sydney',
    phone: '123456789',
    email: 'sydney@gmail.com',
    image: {
      url: 'https://images.unsplash.com/photo-1533738363-b7f9aef128ce?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8M3x8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=800&q=60',
      alt: 'cute-cat',
    },
    favoured: false,
    color: 'grey',
    gender: 'female',
  };

  it('Should display the name of the cat', () => {
    render(<Card {...cat} index={1} />);
    expect(
      screen.getByRole('heading', {
        name: /sydney/i,
      })
    ).toBeInTheDocument();
  });

  it('Should display the phone number of the owner', () => {
    render(<Card {...cat} index={1} />);
    expect(screen.getByText(cat.phone)).toBeInTheDocument();
  });

  it('Should display the email of the owner', () => {
    render(<Card {...cat} index={1} />);
    expect(screen.getByText(cat.email)).toBeInTheDocument();
  });

  it('Should show the image of the cat with correct src', () => {
    render(<Card {...cat} index={1} />);
    expect(screen.getByAltText(cat.image.alt)).toBeInTheDocument();
    expect((screen.getByAltText(cat.image.alt) as HTMLImageElement).src).toBe(cat.image.url);
  });

  it('Should show outlined heart', () => {
    render(<Card {...cat} index={1} />);
    expect(screen.queryByTestId('filled-heart')).not.toBeInTheDocument();
    expect(screen.getByTestId('outlined-heart')).toBeInTheDocument();
  });

  it('Should show filled heart', () => {
    render(<Card {...cat} index={1} favoured={true} />);
    expect(screen.queryByTestId('outlined-heart')).not.toBeInTheDocument();
    expect(screen.getByTestId('filled-heart')).toBeInTheDocument();
  });

  it('Should toggle heart status', () => {
    render(<Card {...cat} index={1} />);

    userEvent.click(screen.getByRole('button'));
    expect(screen.getByTestId('filled-heart')).toBeInTheDocument();
    expect(screen.queryByTestId('outlined-heart')).not.toBeInTheDocument();

    userEvent.click(screen.getByRole('button'));
    expect(screen.queryByTestId('filled-heart')).not.toBeInTheDocument();
    expect(screen.getByTestId('outlined-heart')).toBeInTheDocument();
  });
});
