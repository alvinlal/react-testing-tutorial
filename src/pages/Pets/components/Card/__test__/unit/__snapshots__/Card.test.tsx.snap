// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.tsx Match's snapshot 1`] = `
<div
  className="card"
  data-testid="cat-card"
>
  <div
    className="card-header"
  >
    <img
      alt="cute-cat"
      className="card-img"
      src="https://images.unsplash.com/photo-1533738363-b7f9aef128ce?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8M3x8Y2F0fGVufDB8fDB8fA%3D%3D&auto=format&fit=crop&w=800&q=60"
    />
    <button
      className="heart"
      onClick={[Function]}
    >
      <svg
        data-testid="outlined-heart"
      >
        heartOutlined.svg
      </svg>
    </button>
    <div
      className="card-content"
    >
      <h3>
        Sydney
      </h3>
      <p>
        123456789
      </p>
      <p>
        sydney@gmail.com
      </p>
    </div>
  </div>
</div>
`;
